contracts/services/reverse-splitter/src/msg.rs

```diff
@@ -0,0 +1,432 @@
+        !self.eq(other)
+    }
+}
+
+impl OptionalServiceConfig {
+    pub fn update_config(self, deps: &DepsMut, config: &mut Config) -> Result<(), ServiceError> {
+        // First update output_addr & base_denom (if needed)
+        if let Some(output_addr) = self.output_addr {
+            config.output_addr = output_addr.to_addr(deps.api)?;
+        }
+
+        if let Some(base_denom) = self.base_denom.clone() {
+            config.base_denom = base_denom
+                .into_checked(deps.as_ref())
+                .map_err(|err| ServiceError::ConfigurationError(err.to_string()))?;
+        }
+
+        // Then validate & update splits (if needed)
+        if let Some(splits) = self.splits {
+            validate_splits(
+                deps.api,
+                &splits,
+                // Use the new base_denom if it was updated, or the existing one (as unchecked)
+                &self.base_denom.unwrap_or(match &config.base_denom {
+                    CheckedDenom::Native(denom) => UncheckedDenom::Native(denom.clone()),
+                    CheckedDenom::Cw20(addr) => UncheckedDenom::Cw20(addr.to_string()),
+                }),
+            )?;
+
+            config.splits = convert_to_checked_configs(deps.as_ref(), &splits)?;
+        }
+        Ok(())
+    }
+}
+
+fn validate_splits(
+    api: &dyn cosmwasm_std::Api,
+    splits: &Vec<UncheckedSplitConfig>,
+    base_denom: &UncheckedDenom,
+) -> Result<(), ServiceError> {
+    if splits.is_empty() {
+        return Err(ServiceError::ConfigurationError(
+            "No split configuration provided.".to_string(),
+        ));
+    }
+
+    let mut denom_map: HashMap<String, ()> = HashMap::new();
+    for split in splits {
+        split.account.to_addr(api)?;
+        // Note: can't validate denom without the deps
+
+        // Ensure splits are unique in split configs
+        let key = format!("{:?}/{:?}", split.account, split.denom);
+        if denom_map.contains_key(&key) {
+            return Err(ServiceError::ConfigurationError(format!(
+                "Duplicate split '{:?}/{:?}' in split config.",
+                split.account, split.denom
+            )));
+        }
+        denom_map.insert(key, ());
+
+        match (split.amount, &split.ratio) {
+            (Some(_), None) => {
+                if split.factor.is_some() {
+                    return Err(ServiceError::ConfigurationError(
+                        "Invalid split config: a factor cannot be specified with an amount."
+                            .to_string(),
+                    ));
+                }
+            }
+            (None, Some(UncheckedRatioConfig::FixedRatio(ratio))) => {
+                if split.denom == *base_denom && *ratio != Decimal::one() {
+                    return Err(ServiceError::ConfigurationError(
+                        "Invalid split config: fixed ratio for base denom must be 1.".to_string(),
+                    ));
+                }
```

Comment:
Just thought about this and probably applies to splitter too. Is it worth validating ratios so that they are 0 < ratio <= 1?

Reply:
I've added a test to check for non-zero amount & ratio. For the reverse splitter, the ratio doesn't have to be between 0 and 1 (it can be an exchange rate for the base denom). For the Splitter, the check on the sum of ratios being 1 is sufficient.